crates/refunder/src/submitter.rs

```diff
@@ -0,0 +1,227 @@
+use anyhow::{anyhow, Result};
+use contracts::CoWSwapEthFlow;
+use database::OrderUid;
+use ethcontract::{
+    transaction::{confirm::ConfirmParams, ResolveCondition},
+    Account, GasPrice, U256,
+};
+use gas_estimation::{GasPrice1559, GasPriceEstimating};
+use shared::{
+    ethrpc::Web3,
+    submitter_constants::{TX_ALREADY_KNOWN, TX_ALREADY_MINED},
+};
+
+// Max gas used for submitting transactions
+// If the gas price is higher than this value,
+// the service will temproarily not refund users
+const MAX_GAS_PRICE: u64 = 500_000_000_000u64;
+
+// The gas price buffer determine the gas price buffer used to
+// send out EIP1559 txs.
+// Example: If the prevailing gas is 10Gwei and the buffer is 20
+// then the gas_price used will be 12.
+const GAS_PRICE_BUFFER_IN_PERCENT: f64 = 30.0f64;
+// Max priority fee that the refunder is willing to pay. (=2 Gwei)
+const MAX_PRIORITY_FEE_TIP: u64 = 2_000_000_000u64;
+// In order to resubmit a new tx with the same nonce, the gas price needs to
+// be increased by at least 10 percent. We increase it by 12 percent
+const GAS_PRICE_BUMP: f64 = 1.12f64;
+
+pub struct Submitter {
+    pub web3: Web3,
+    pub ethflow_contract: CoWSwapEthFlow,
+    pub account: Account,
+    pub gas_estimator: Box<dyn GasPriceEstimating>,
+    pub gas_price_of_last_submission: Option<U256>,
+    pub nonce_of_last_submission: Option<U256>,
+}
+
+impl Submitter {
+    async fn get_submission_nonce(&self) -> Result<U256> {
+        // this command returns the tx count ever mined at the latest block
+        // Mempool tx are not considered.
+        self.web3
+            .eth()
+            .transaction_count(self.account.address(), None)
+            .await
+            .map_err(|err| anyhow!("Could not get latest nonce due to err: {:}", err))
+    }
+
+    pub async fn submit(
+        &mut self,
+        uids: Vec<OrderUid>,
+        encoded_ethflow_orders: Vec<EncodedEthflowOrder>,
+    ) -> Result<()> {
+        let confirm_params = ConfirmParams {
+            block_timeout: Some(5),
+            ..Default::default()
+        };
+        let resolve_conditions = ResolveCondition::Confirmed(confirm_params);
+        let gas_price_estimation = self.gas_estimator.estimate().await?;
+        let nonce = self.get_submission_nonce().await?;
+        let gas_price = calculate_submission_gas_price(
+            self.gas_price_of_last_submission,
+            gas_price_estimation,
+            nonce,
+            self.nonce_of_last_submission,
+        )?;
+        let max_fee_per_gas = match gas_price {
+            GasPrice::Eip1559 {
+                max_fee_per_gas, ..
+            } => max_fee_per_gas,
+            _ => {
+                return Err(anyhow!(
+                    "Unreachable state while accessing the max_fee_per_gas"
+                ))
+            }
+        };
+
+        // Gas prices are capped at MAX_GAS_PRICE
+        if max_fee_per_gas < U256::from(MAX_GAS_PRICE) {
```

Comment:
To prevent the code drifting to the right you could return early here instead of having so much code in the `if` block.